<b><h1>DNS Water Torture σε Recursive DNS Servers</h1></b>

<p align="justify">Στην άσκηση αυτή, θα αναπτύξετε ένα μηχανισμό αντιμετώπισης επιθέσεων <i>DNS Water Torture</i> σε <i>Recursive DNS Servers</i> (δίνεται <a href="https://github.com/nkostopoulos/StochasticsLabPublic/blob/master/lab9/Lab_theory.pdf">υποστηρικτικό υλικό</a> για την άσκηση). Ουσιαστικά, το πρόβλημα αυτό είναι ένα πρόβλημα text classification που ζητάει το διαχωρισμό ονομάτων DNS σε <i>έγκυρα (valid)</i> και <i>άκυρα (invalid)</i>. </p>

<p align="justify">Με τη βοήθεια του αλγορίθμου Naive Bayes Classifier, θα διαχωρίσετε τα prefixes των <i>ονομάτων DNS</i> σε <i>έγκυρα</i> και <i>άκυρα</i>. Στην άσκηση αυτή, ως <i>prefix</i> ορίζουμε το <i>πρώτο label</i> ενός <i>ονόματος DNS</i>. Για παράδειγμα, το prefix του ονόματος www.ntua.gr είναι το <i>www</i>, ενώ το prefix του ονόματος dolly.netmode.ece.ntua.gr είναι το <i>dolly</i>.</p>

<p align="justify">Ο αλγόριθμος θα εκπαιδευτεί με <u>έγκυρα ονόματα</u> (αρχείο <a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab9/valid_training.txt">valid_training.txt</a>) και άκυρα ονόματα (αρχείο <a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab9/invalid_training.txt">invalid_training.txt</a>). Η δοκιμή του θα γίνει στο <u>test set</u>, που περιλαμβάνει, επίσης, έγκυρα (αρχείο <a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab9/valid_test.txt">valid_test.txt</a>) και άκυρα (αρχείο <a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab9/invalid_test.txt">invalid_test.txt</a>) ονόμάτα.</p>

<p align="justify">Να απαντήσετε στις ακόλουθες ερωτήσεις:</p>
<ul>
<li>Ποια είναι η παραδοχή του αλγορίθμου <b><i>Naive Bayes Classifier</i></b>; Ποια είναι τα πλεονεκτήματα του αλγορίθμου;</li>
<li>Να περιγράψετε σύντομα τη λειτουργία του αλγορίθμου <b><i>Naive Bayes Classifier</i></b>.</li>
<li>Μελετήστε τα prefixes που περιλαμβάνονται στα αρχεία <a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab9/valid_training.txt">valid_training.txt</a> και <a href="https://raw.githubusercontent.com/nkostopoulos/StochasticsLabPublic/master/lab9/invalid_training.txt">invalid_training.txt</a>. Σχολιάστε μερικές από τις βασικές διαφορές ανάμεσα στα prefixes των δύο αρχείων που θα μπορούσαν να φανούν χρήσιμες για το διαχωρισμό των έγκυρων και των άκυρων ονομάτων.</li>
<li>Ποια είναι τα 7 features που έχουν επιλεχθεί στον κώδικα της άσκησης;</li>
<li>Να εκτελέσετε τον κώδικα της άσκησης. Καταγράψετε και σχολιάστε την ακρίβεια του αλγορίθμου πάνω στο <i>test set</i> για τα valid και τα invalid names ξεχωριστά. Ποιο είναι το μέγεθος του <i>training set</i>; Πόση ώρα διήρκησε η εκπαίδευση του αλγορίθμου;</li>
<li>Αφού εκτελέσετε τον κώδικα, θα παρατηρήσετε πως έχουν παραχθεί δύο αρχεία: <i>problematic_valid.txt</i> και <i>problematic_invalid.txt</i>. Μελετώντας τον κώδικα, γιατί πιστεύετε ότι τα prefixes που περιλαμβάνονται σε αυτά τα αρχεία δημιουργούν πρόβλημα στην περίπτωσή μας; Θυμηθείτε τα μειονεκτήματα του αλγορίθμου <b><i>Naive Bayes Classifier</i></b>. Τι θα μπορούσατε να κάνετε για να λύσετε το πρόβλημα αυτό (<a href="https://www.cs.cmu.edu/~tom/mlbook/NBayesLogReg.pdf">https://www.cs.cmu.edu/~tom/mlbook/NBayesLogReg.pdf</a>);</li>
<li>Μελετώντας τη συνάρτηση <i>find_prob()</i>, θα δείτε πως λείπουν οι πιθανότητες <i>prior</i> από τους υπολογισμούς. Τι παραδοχή έχουμε κάνει για τις <i>prior</i> πιθανότητες και κατά συνέπεια για τις δύο κατηγορίες ταξινόμησης (<i>valid</i>, <i>invalid</i>) στην άσκησή μας; Πώς αλλιώς θα μπορούσατε να τις επιλέξετε;</li>
<li>Μπορείτε να προτείνετε κάποιο επιπρόσθετο <i>feature</i> για τον αλγόριθμο; (δε χρειάζεται να το υλοποιήσετε)</li>
</ul>

<p align="justify"><b>Πηγές Δεδομένων</b></p>
<ul>
<li>Valid ονόματα: <a href="https://www.kaggle.com/cheedcheed/top1m">https://www.kaggle.com/cheedcheed/top1m</a> (πολλά από τα κορυφαία σε επισκεψιμότητα site)</li>
<li>Invalid ονόματα: Παράχθηκαν με το πρόγραμμα <a href="https://github.com/nkostopoulos/StochasticsLabPublic/blob/master/lab9/generator.py">generator.py</a> που θα βρείτε μαζί με τα υπόλοιπα αρχεία της άσκησης.</li>
</ul>


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
%%time
problematic1 = open("problematic_valid.txt", "w")
problematic2 = open("problematic_invalid.txt", "w")

def load_file(file_name):
    fd = open(file_name, "r")
    my_set = set()
    for prefix in fd:
        prefix = prefix.rstrip()
        my_set.add(prefix)
    return my_set

def calculate_probabilities(dataset):
    stats = dict()
    for index in range(0, 7):
        stats[index] = dict()
    for prefix in dataset:
        features = handle_name(prefix)
        for index in range(0, 7):
            try:
                stats[index][features[index]] += 1
            except:
                stats[index][features[index]] = 1

    dataset_size = len(dataset)
    for index in range(0, 7):
        for key in stats[index]:
            stats[index][key] /= dataset_size
    return stats

def handle_name(prefix):
    total_length = len(prefix)
    total_digits, max_numeric_sequence = numeric(prefix)
    total_consonants, max_consonants_sequence = consonants(prefix)
    total_vowels, max_vowels_sequence = vowels(prefix)
    return total_length, total_digits, max_numeric_sequence, total_consonants, max_consonants_sequence, total_vowels, max_vowels_sequence

def vowels(prefix):
    total_vowels = 0
    vowels_sequence = list()
    current_sequence = 0
    for char in prefix:
        if char == 'a' or char == 'e' or char == 'i' or char == 'o' or char == 'u':
            total_vowels += 1
            current_sequence += 1
        else:
            vowels_sequence.append(current_sequence)
            current_sequence = 0
    vowels_sequence.append(current_sequence)
    max_vowels_sequence = max(vowels_sequence)
    return total_vowels, max_vowels_sequence

def consonants(prefix):
    total_consonants = 0
    consonants_sequence = list()
    current_sequence = 0
    for char in prefix:
        if char != 'a' and char != 'e' and char != 'i' and char != 'o' and char != 'u' and char != '-' and char.isdigit() == False:
            total_consonants += 1
            current_sequence += 1
        else:
            consonants_sequence.append(current_sequence)
            current_sequence = 0
    consonants_sequence.append(current_sequence)
    max_consonants_sequence = max(consonants_sequence)
    return total_consonants, max_consonants_sequence

def numeric(prefix):
    total_digits = 0
    numeric_sequence = list()
    current_sequence = 0
    for char in prefix:
        if char.isdigit() == True:
            total_digits += 1
            current_sequence += 1
        else:
            numeric_sequence.append(current_sequence)
            current_sequence = 0
    numeric_sequence.append(current_sequence)
    max_numeric_sequence = max(numeric_sequence)
    return total_digits, max_numeric_sequence

def find_prob(prefix, stats, fd):
    tl, td, mns, tc, mcs, tv, mvs = handle_name(prefix)
    try:
        prob = stats[0][tl] * stats[1][td] * stats[2][mns] * stats[3][tc] * stats[4][mcs] * stats[5][tv] * stats[6][mvs]
    except:
        prob = 0
        fd.write(prefix + "\n")
    return prob

def apply_on_test_set(test_set, category, valid_stats, invalid_stats, fd):
    misclassifications = 0
    names_processed = 0
    for prefix in test_set:
        valid_prob = find_prob(prefix, valid_stats, fd)
        invalid_prob = find_prob(prefix, invalid_stats, fd)
        if valid_prob != 0 and invalid_prob != 0:
            names_processed += 1
            if category == "valid" and valid_prob < invalid_prob:
                misclassifications += 1
            elif category == "invalid" and valid_prob > invalid_prob:
                misclassifications += 1
    return misclassifications, names_processed

if __name__ == "__main__":
    # Load valid names training set
    valid_names_training = load_file("./valid_training.txt")
    # Load valid names test set
    valid_names_test = load_file("./valid_test.txt")
    # Load invalid names training set
    invalid_names_training = load_file("./invalid_training.txt")
    # Load invalid names test set
    invalid_names_test = load_file("./invalid_test.txt")

    valid_stats = calculate_probabilities(valid_names_training)
    invalid_stats = calculate_probabilities(invalid_names_training)

    valid_misclassifications, valid_names_processed = apply_on_test_set(valid_names_test, "valid", valid_stats, invalid_stats, problematic1)
    invalid_misclassifications, invalid_names_processed = apply_on_test_set(invalid_names_test, "invalid", valid_stats, invalid_stats, problematic2)

    print("Valid names misclassified as invalid - Ratio: ", (valid_misclassifications / valid_names_processed) * 100)
    print("Invalid names misclassified as valid - Ratio: ", (invalid_misclassifications / invalid_names_processed) * 100)


Valid names misclassified as invalid - Ratio:  1.9172947402180884
Invalid names misclassified as valid - Ratio:  1.2655821677971868
CPU times: user 30.6 s, sys: 207 ms, total: 30.8 s
Wall time: 32.3 s


In [8]:
print(valid_stats)

{0: {13: 0.060977142857142855, 6: 0.07845857142857143, 16: 0.03039714285714286, 8: 0.10063142857142857, 10: 0.09508714285714286, 5: 0.05750285714285714, 9: 0.10058, 7: 0.09083285714285715, 11: 0.08423571428571429, 15: 0.040222857142857144, 19: 0.01268857142857143, 21: 0.006614285714285714, 14: 0.050147142857142855, 4: 0.03768571428571429, 12: 0.07327142857142857, 17: 0.023665714285714286, 22: 0.004825714285714286, 3: 0.013177142857142858, 18: 0.017365714285714286, 24: 0.002354285714285714, 20: 0.009247142857142857, 25: 0.0016685714285714285, 23: 0.0032485714285714284, 29: 0.0003928571428571429, 27: 0.0008242857142857142, 28: 0.0005857142857142858, 26: 0.0011357142857142857, 30: 0.0002985714285714286, 2: 0.0009414285714285715, 32: 0.00016714285714285713, 31: 0.00024142857142857142, 39: 2.1428571428571428e-05, 37: 4.428571428571428e-05, 45: 5.7142857142857145e-06, 42: 5.7142857142857145e-06, 35: 6.857142857142857e-05, 33: 0.00010714285714285714, 36: 5.142857142857143e-05, 57: 1.428571428

In [7]:
%time
import pandas as pd
import numpy as np
valid_train = pd.read_csv('valid_training.txt', header=None)
invalid_train = pd.read_csv('invalid_training.txt', header=None)

mean_valid_train = int(valid_train[0].str.len().mean())
mean_invalid_train = int(invalid_train[0].str.len().mean())

print(f'Το μέσο μήκος συμβολοσειράς για το αρχείο valid_training.txt είναι {mean_valid_train}.')
print(f'Το μέσο μήκος συμβολοσειράς για το αρχείο invalid_training.txt είναι {mean_valid_train}.')

CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 5.01 µs
Το μέσο μήκος συμβολοσειράς για το αρχείο valid_training.txt είναι 10.
Το μέσο μήκος συμβολοσειράς για το αρχείο invalid_training.txt είναι 10.


Θα εκτελέσουμε τον παρακάτω κώδικα για να υπολογίσουμε την ακρίβεια του αλγορίθμου πάνω στο test set για τα valid και τα invalid names

In [10]:
%%time
if __name__ == "__main__":
    # Load valid names training set
    valid_names_training = load_file("./valid_training.txt")
    # Load valid names test set
    valid_names_test = load_file("./valid_test.txt")
    # Load invalid names training set
    invalid_names_training = load_file("./invalid_training.txt")
    # Load invalid names test set
    invalid_names_test = load_file("./invalid_test.txt")

    test_valid_stats = calculate_probabilities(valid_names_test)
    test_invalid_stats = calculate_probabilities(invalid_names_test)

    test_valid_misclassifications, test_valid_names_processed = apply_on_test_set(valid_names_test, "valid", test_valid_stats, test_invalid_stats, problematic1)
    test_invalid_misclassifications, test_invalid_names_processed = apply_on_test_set(invalid_names_test, "invalid", test_valid_stats, test_invalid_stats, problematic2)

    print("Valid names misclassified as invalid - Ratio: ", (test_valid_misclassifications / test_valid_names_processed) * 100)
    print("Invalid names misclassified as valid - Ratio: ", (test_invalid_misclassifications / test_invalid_names_processed) * 100)

Valid names misclassified as invalid - Ratio:  3.726564753389097
Invalid names misclassified as valid - Ratio:  1.2421115897024941
CPU times: user 10.6 s, sys: 211 ms, total: 10.8 s
Wall time: 10.9 s


ΑΠΑΝΤΗΣΕΙΣ:

i)

Βασική Παραδοχή Naive Bayes Classifier:


Στη μηχανική μάθηση, οι απλοί ταξινομητές Bayes είναι μία οικογένεια απλών <<πιθανολογικών ταξινομητών>> που βασίζονται στην εφαρμογή του θεωρήματος Bayes με ισχυρές (ή και αφελής) υποθέσεις , ανεξάρτητες μεταξύ των features(χαρακτηριστικών).

Συνεπώς, η βασική παραδοχή του αλγορίθμου Naive Bayes Classifier είναι η υπόθεση της ανεξαρτησίας των χαρακτηριστικών. Συγκεκριμένα, ο αλγόριθμος υποθέτει ότι κάθε χαρακτηριστικό σε ένα δείγμα είναι ανεξάρτητο από τα υπόλοιπα χαρακτηριστικά, δεδομένης της κλάσης στην οποία ανήκει το δείγμα. Αυτή η παραδοχή είναι αρκετά ισχυρή και συχνά μη ρεαλιστική, καθώς τα χαρακτηριστικά στις πραγματικές εφαρμογές συχνά αλληλοεξαρτώνται. Παρ' όλα αυτά, ο Naive Bayes Classifier λειτουργεί αρκετά καλά σε πολλές περιπτώσεις , ακόμα και όταν η παραδοχή της ανεξαρτησίας παραβιάζεται.

Για παράδειγμα, στον Naive Bayes , για ένα πρόβλημα κατηγοριοποίησης κειμένου, η πιθανότητα ενός κειμένου να ανήκει σε μία συγκεκριμένη κατηγορία υπολογίζεται με βάση την υπόθεση ότι η παρουσία ή απουσία μίας λέξης είναι ανεξάρτητη από την παρουσία ή την απουσία άλλων λέξεων, δεδομένης της κατηγορίας.

Μαθηματικά η παραδοχή αυτή διατυπώνεται ως εξής:

P(X1,X2,...,Xn|C)= P(X1|C)* P(X2|C)...*P(Xn|C), όπου

x1,x2,...,xn είναι τα χαρακτηριστικά του δείγματος

C είναι η κατηγορία στην οποία ανήκει το δείγμα

P(Xi|C) είναι η πιθανότητα του χαρακτηριστικού Xi δεδομένης της κατηγορίας C.


Πλεονεκτήματα Naive Bayes Classifier:

1. Εφαρμόζεται εύκολα και είναι υπολογιστικά αποδοτικός
2. Αποτελεσματικός σε περιπτώσεις με μεγάλο αριθμό από features
3. Αποδίδει πολύ καλά ακόμη και όταν έχουμε περιορισμένα training data
4. Αντιδράει καλά και αποδίδει αποτελεσματικά στην παρουσία κατηγορηματικών features
5. Στον αλγόριθμο αυτό, τα features με αριθμιτικά δεδομένα προέρχονται από κανονική κατανομή.

Εφαρμογές Naive Bayes Classifier:

1. Spam Email Filtering: Χωρίζει τα email σε spam ή non-spam με βάση τα features
2. Text Classification: Χρησιμοποιείται για ανάλυση συναισθήματος στα κείμενα, κατηγοριοποίησης document και διαχωρισμό θεμάτων.
3. Medical Diagnosis: Βοηθάει στη πρόβλεψη της πιθανότητας κάποιας ασθένειας με βάση κάποια συμπτώματα
4. Weather Prediction: Κατηγοριοποιεί τις καιρικές συνθήκες με βάση διάφορους παράγοντες
5. Credit Scoring: Μπορεί να αξιολογήσει τη πιστοληπτική ικανότητα των ατόμων για την έγκριση του δανείου.

ii)

Το θεώρημα Bayes βρίσκει τη πιθανότητα να συμβεί ένα γεγονός δεδομένης της πιθανότητας ενός άλλου γεγονότος που έχει ήδη συμβεί. Το θεώρημα του Bayes δηλώνεται μαθηματικά ως η ακόλουθη εξίσωση:

P(A|B)=P(B|A)*P(A)/P(B), όπου Α και Β είναι γεγονότα και P(B)!=0

Στην ουσία προσπαθούμε να βρούμε τη πιθανότητα του γεγονότος Α , δεδομένου ότι το γεγονός Β είναι αληθές. Το συμβάν Β ορίζεται επίσης ως αποδεικτικό στοιχείο.

Το P(A) είναι η προηγούμενη πιθανότητα, δηλαδή η πιθανότητα πριν εμφανιστεί το αποδεικτικό στοιχείο. Η απόδειξη είναι μία τιμή χαρακτηριστικού ενός αγνώστου στιγμιότυπου (εδώ είναι το συμβάν Β).

Το P(B) είναι η οριακή πιθανότητα. Πιθανότητα Απόδειξης

Το P(A|B) είναι μία εκ των υστέρων πιθανότητα του Β, δηάδή η πιθανότητα συμβάντος μετά την εμφάνιση στοιχείων.

Το P(B|A) είναι η πιθανότητα η υπόθεση να βγει αληθινή με βάση το αποδεικτικό στοιχείο.

Γενικά έχουμε ότι:
1. Συλλέγουμε τα δεδομένα εκπαίδευσης και τα χωρίζουμε σε διαφορετικές κλάσεις
2. Υπολογίζουμε τις απαραίτητες πιθανότητες από τα δεδομένα εκπαίδευσης
3. Για ένα νέο δείγμα, υπολογίζουμε τη πιθανότητα για κάθε κλάση
4. Κατατάσουμε το δείγμα στη κλάση με τη μεγαλύτερη πιθανότητα


Εκπαίδευση:

Υπολογίζουμε τις αρχικές πιθανότητες P(C) για κάθε κατηγορία C.
Υπολογίζουμε τις πιθανότητες P(Xi∣C) για κάθε χαρακτηριστικό Xi δεδομένης της κατηγορίας C.
Ταξινόμηση:

Για ένα νέο δείγμα χαρακτηριστικών 𝑋={𝑋1,𝑋2,...,𝑋𝑛}
, υπολογίζουμε την πιθανότητα κάθε κατηγορίας C με βάση το προϊόν των πιθανοτήτων των χαρακτηριστικών: 𝑃(𝐶∣𝑋)∝𝑃(𝐶)⋅𝑃(𝑋1∣𝐶)⋅𝑃(𝑋2∣𝐶)⋅.. (𝑋𝑛∣𝐶). Επιλέγουμε την κατηγορία με τη μέγιστη πιθανότητα.

Παραδοχές Naive Bayes:
1. Ανεξαρτησία χαρακτηριστικών: Τα χαρακτηριστικά (features) των δεδομένων είναι υπό όρους ανεξάρτητα μεταξύ τους, δεδομένου της ετικέτας κλάσης (label class)
2. Τα συνεχή χαρακτηριστικά κατανέμονται κανονικά: Εάν ένα χαρακτηριστικό είναι συνεχές , τότε θεωρείται ότι είναι κανονικά κατανεμημένο σε κάθε κλάση
3. Τα διακριτά χαρακτηριστικά έχουν πολυωνυμικές κατανομές: Εάν ένα χαρακτηριστικό είναι διακριτό , τότε θεωρείται ότι έχει πολυωνυμική κατανομή
σε κάθε κλάση
4. Όλα τα χαρακτηριστικά είναι εξίσου σημαντικά, δηλαδή συμβάλουν εξίσου στη πρόβλεψη του label class (y)
5. Δεν λείπουν δεδομένα: Τα δεδομένα δεν πρέπει να περιέχουν τιμές που λείπουν.

iii)

In [19]:
from collections import Counter


valid_training = 'valid_training.txt'
invalid_training = 'invalid_training.txt'

# Function to read lines from a file
def read_lines(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        return file.readlines()

# Function to extract prefixes from lines
def extract_prefixes(lines, prefix_length=3):
    return [line[:prefix_length] for line in lines]


# Function to count prefix frequencies
def count_prefixes(prefixes):
    return Counter(prefixes)



valid_lines = read_lines(valid_training)
invalid_lines = read_lines(invalid_training)


prefix_length = 3
valid_prefixes = extract_prefixes(valid_lines, prefix_length)
invalid_prefixes = extract_prefixes(invalid_lines, prefix_length)

# Count prefix frequencies
valid_prefix_counter = count_prefixes(valid_prefixes)
invalid_prefix_counter = count_prefixes(invalid_prefixes)


print("Most common prefixes in valid dataset")
counter=0
valid_prefix_counter={k: v for k, v in sorted(valid_prefix_counter.items(), key=lambda item: item[1],reverse=True)}
for prefix in valid_prefix_counter:
  print(f"{prefix}:{valid_prefix_counter[prefix]}")
  counter+=1
  if (counter==10):
    break
print("\n\n")

print("Most common prefixes in invalid dataset")
c=0
invalid_prefix_counter={k: v for k, v in sorted(invalid_prefix_counter.items(), key=lambda item: item[1],reverse=True)}
for prefix in invalid_prefix_counter:
  print(f"{prefix}:{invalid_prefix_counter[prefix]}")
  c+=1
  if (c==10):
    break

Most common prefixes in valid dataset
the:8300
pro:3931
fre:2818
tra:2734
por:2496
con:2410
com:2303
mar:2276
car:2269
sta:2189



Most common prefixes in invalid dataset
q8i:31
8o9:30
2yk:30
clp:30
kyi:30
96f:30
3y4:30
sr0:30
1cn:30
rpn:29


iv)

Τα 7 features που χρησιμοποιούνται είναι:

1. total_length
2. total_digits
3. total_consonats
4. total_vowels
5. max_numeric_sequence
6. max_consonats_sequence
7. max_vowels_sequence

Το  total length αντιπροσοπεύει το μήκος κάθε συμβολοσειράς.

Το total digits αντικατοπτρίζει το σύνολο των χαρακτήρων (strings) της συμβολοσειράς που αντιστοιχούν σε ψηφία (0,1,2,3,4,5,6,7,8,9).

Το total_consonats δείχνει το σύνολο των συμφώνων της συμβολοσειράς

Το total_vowels δείχνει το σύνολο των φωνήεν της συμβολοσειράς

Το max_numeric_sequence δείχνει το μέγιστο πλήθος ψηφίων που εμφανίζονται σε μία συμβολοσειρά

Το max_consonats_sequence δείχνει το μέγιστο πλήθος συμφώνων που εμφανίζονται σε κάποια συμβολοσειρά του δείγματος

Το max_vowel_sequence δείχνει το μέγιστο πλήθος φωνήεν που εμφανίζονται σε κάποια συμβολοσειρά του δείγματος


v)

Χρησιμοποιώντας την εντολή %%time βλέπουμε ότι ο χρόνος εκπαίδευσης του αλγορίθμου είναι:

CPU times: user 10.6 s, sys: 211 ms, total: 10.8 s

Wall time: 10.9 s

Τόσο όσον αφορα την ακρίβεια του αλγορίθμου καταλήγουμε στα εξής συμπεράσματα:

Valid names misclassified as invalid - Ratio:  3.726564753389097


Η πολύ υψηλή τιμή αυτής της αναλογίας μας προβληματίζει διότι αυτό σημαίνει   ότι το σύστημα  απορρίπτει πολλά έγκυρα ονόματα. Αυτό μπορεί να έχει αρνητικές επιπτώσεις, διότι χάνουμε ένα αρκετά μεγάλο αριθμό ονομάτων ως μη έγκυρα.

Invalid names misclassified as valid - Ratio:  1.2421115897024941

Αυτή η τιμή υποδηλώνει ότι το σύστημα κάνει επίσης λάθη αποδεχόμενο άκυρα ονόματα, αν και όχι τόσο συχνά όσο απορρίπτει έγκυρα ονόματα. Αυτό μπορεί να οδηγήσει σε προβλήματα με την ακεραιότητα των δεδομένων και την ασφάλεια.


Το μέγεθος του training set είναι:

In [11]:
print(len(valid_names_training)+len(invalid_names_training))

1400000


vi)

Τα μειονεκτήματα του αλγορίθμου Naive Bayes είναι τα εξής:

1. Ο αλγόριθμος υποθέτει ότι τα χαρακτηριστικά είναι ανεξάρτητα, τα οποία μπορεί να μην ισχύουν πάντα σε δεδομένα πραγματικού κόσμου.
2. Μπορεί να επηρεαστεί από άσχετα χαρακτηριστικά.
3. Μπορεί να εκχωρήσει μηδενική πιθανότητα σε μη ορατά γεγονότα, οδηγώντας σε κακή γενίκευση.

Τα 2 αρχεία που δημιουργούνται (problematic_valid.txt και problematic_invalid.txt) δημιουργούν το πρόβλημα της μηδενικής συχνότητας (zero frequency problem). Το πρόβλημα αυτό προκύπτει κατά τη χρήση του ταξινομητή Naive Bayes όταν ένα συγκεκριμένο feature (χαρακτηριστικό) δεν εμφανίζεται σε κανένα παράδειγμα εκπαίδευσης για μια δεδομένη κατηγορία. Αυτό σημαίνει ότι η πιθανότητα εμφάνισης αυτού του feature για την κατηγορία αυτή θα είναι μηδενική, κάτι που μπορεί να οδηγήσει σε ανεπιθύμητα αποτελέσματα κατά την ταξινόμηση νέων δεδομένων.

Πιο συγκεκριμένα, στον ταξινομητή Naive Bayes, οι πιθανότητες των χαρακτηριστικών πολλαπλασιάζονται για να υπολογιστεί η συνολική πιθανότητα μιας κατηγορίας. Εάν οποιοδήποτε χαρακτηριστικό έχει πιθανότητα μηδέν, η συνολική πιθανότητα για αυτή την κατηγορία θα είναι επίσης μηδέν, ανεξάρτητα από τις άλλες πιθανότητες. Αυτό μπορεί να οδηγήσει σε λανθασμένες ταξινομήσεις, καθώς η κατηγορία με πιθανότητα μηδέν θα απορριφθεί αυτόματα.


Μια προσέγγιση για να ξεπεραστεί αυτό το «πρόβλημα μηδενικής συχνότητας» είναι να προσθέσουμε +1 στο πλήθος για κάθε συνδυασμό τιμής χαρακτηριστικού-κλάσης όταν η τιμή ενος feature δεν εμφανίζεται με κάθε τιμή της κλάσης.

Αυτό θα οδηγήσει στην αφαίρεση όλων των μηδενικών τιμών από τις κλάσεις και, ταυτόχρονα, δεν θα επηρεάσει τη συνολική σχετική συχνότητα των κλάσεων.

Αυτή η διαδικασία «εξομάλυνσης» των δεδομένων μας με την προσθήκη ενός αριθμού είναι γνωστή ως additive smoothing (εξομάλυνση προσθέτων) , που ονομάζεται επίσης εξομάλυνση Laplace (Laplace smoothing).

vii)


Η συνάρτηση find_prob()  στον ταξινομητή Naive Bayes βασίζεται σε πιθανότητες που υπολογίζονται από τα δεδομένα εκπαίδευσης. Αν δεν λαμβάνουμε υπόψη τις prior πιθανότητες, τότε αυτό σημαίνει ότι έχουμε κάνει την παραδοχή ότι οι δύο κατηγορίες (valid και invalid) έχουν ίσες πιθανότητες εμφάνισης. Αυτή είναι μια απλοποίηση που μπορεί να μην ισχύει πάντα στην πραγματικότητα.


Η παραδοχή λοιπόν, είναι ότι οι prior πιθανότητες είναι ίσες ( 0.5 για κάθε κατηγορία) από τη στιγμή κιόλας που το μέγεθος των δύο συνόλων εκπαίδευσης (ένα για valid και ένα για invalid) είναι κοινό (700.000 συμβολοσειρές έκοστος). Το γεγονός αυτό απλοποιεί τους υπολογισμούς και σημαίνει ότι θεωρούμε πως δεν υπάρχει προκατάληψη υπέρ κάποιας κατηγορίας πριν δούμε τα δεδομένα. Αυτό μπορεί να λειτουργήσει καλά σε περιπτώσεις όπου οι κατηγορίες είναι περίπου ισοκατανεμημένες.

 Υπάρχουν διάφορες εναλλακτικές για την εκτίμηση των prior πιθανοτήτων, μέσω αλγορίθμων που εκτιμούν γενικά παραμέτρους στατιστικών κατανομών, δεδομένου ενός δειγματικού συνόλου. Ο γνωστότερος εξ αυτών είναι μάλλον ο αλγόριθμος Maximum Likelihood Estimation (MLE).

viii)

Ένα επιπλέον feature που θα μπορούσαμε να προσθέσουμε στον αλγόριθμο είναι να υπολογίζει από ποιο string αρχίζουν οι μεγαλύτερες λέξεις και από ποιο οι μικρότερες.

Επίσης ένα άλλο κατά τη γνώμη μου χρήσιμο feature  θα ήταν αν υπολόγιζε ο αλγόριθμος το μέσο μήκος των λέξεων που περιέχουν sting αριθμό(0,1,2,3,4,5,6,7,8,9) και τι μέσο μήκος οι λέξεις που δεν περιέχουν string αριθμό.